### PR TITLE
Add gimmepeers torrent provider

### PR DESCRIPTION
**Pull request by mystycs**
Popular torrent provider which used to be bitsoup.me on the old sickrage.

- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: medusa/providers/torrent/html/gimmepeers.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        self.url = self.urls['base_url']
+
+        self.username = None
+        self.password = None
+        self.minseed = None
+        self.minleech = None
+
+        self.cache = tv.Cache(self)
+
+        self.search_params = {
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: `search_params` need to go in login()

File: medusa/providers/torrent/html/gimmepeers.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            # c20=1&c21=1&c25=1&c24=1&c23=1&c22=1&c1=1
+            'c20': 1, 'c21': 1, 'c25': 1, 'c24': 1, 'c23': 1, 'c22': 1, 'c1': 1
+        }
+
+    def _check_auth(self):
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: these needs to go to bottom

File: medusa/providers/torrent/html/gimmepeers.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        if not self.username or not self.password:
+            logger.log(u'Invalid username or password. Check your settings', logger.WARNING)
+
+        return True
+
+    def login(self):
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: also bottom

File: medusa/providers/torrent/html/gimmepeers.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        if not self.login():
+            return results
+
+        for mode in search_strings:
+
+            logger.log(u'Search Mode: {0}'.format(mode), logger.DEBUG)
+            for search_string in search_strings[mode]:
+                if mode != 'RSS':
+                    logger.log(u'Search string: {0}'.format
+                               (search_string.decode('utf-8')), logger.DEBUG)
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: no need to use decode here

File: medusa/providers/torrent/html/gimmepeers.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    """GimmePeers Torrent provider."""
+
+    def __init__(self):
+
+        TorrentProvider.__init__(self, 'GimmePeers')
+
+        self.urls = {
+            'base_url': 'https://www.gimmepeers.com',
+            'login': 'https://www.gimmepeers.com/takelogin.php',
+            'detail': 'https://www.gimmepeers.com/details.php?id=%s',
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: use {s} and format() instead

File: medusa/providers/torrent/html/gimmepeers.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                    # except (AttributeError, TypeError):
+                    #     continue
+
+                    if not all([title, download_url]):
+                        continue
+
+                        # Filter unseeded torrent
+                    if seeders < self.minseed or leechers < self.minleech:
+                        if mode != 'RSS':
+                            logger.log(u"Discarding torrent because it doesn't meet the minimum seeders or leechers: {0} (S:{1} L:{2})".format
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: See our formatting https://github.com/pymedusa/Medusa/blob/master/medusa/providers/torrent/html/alpharatio.py#L166-L168

File: medusa/providers/torrent/html/gimmepeers.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                                       (title, seeders, leechers), logger.DEBUG)
+                        continue
+
+                    if seeders >= 32768 or leechers >= 32768:
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: What's the point of this?

File: medusa/providers/torrent/html/gimmepeers.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+                    link = cells[1].find('a')
+                    download_url = self.urls['download'] % cells[2].find('a')['href']
+
+                    title = link.getText()
+                    seeders = int(cells[10].getText().replace(',', ''))
+                    leechers = int(cells[11].getText().replace(',', ''))
+                    torrent_size = cells[8].getText()
+                    size = convert_size(torrent_size) or -1
+                    # except (AttributeError, TypeError):
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: This can be removed

File: medusa/providers/torrent/html/gimmepeers.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                    #     continue
+
+                    if not all([title, download_url]):
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: This needs to go right after title and download_url. See https://github.com/pymedusa/Medusa/blob/master/medusa/providers/torrent/html/alpharatio.py#L135-L139

File: medusa/providers/torrent/html/gimmepeers.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                    if seeders < self.minseed or leechers < self.minleech:
+                        if mode != 'RSS':
+                            logger.log(u"Discarding torrent because it doesn't meet the minimum seeders or leechers: {0} (S:{1} L:{2})".format
+                                       (title, seeders, leechers), logger.DEBUG)
+                        continue
+
+                    if seeders >= 32768 or leechers >= 32768:
+                        continue
+
+                    item = {'title': title, 'link': download_url, 'size': size, 'seeders': seeders, 'leechers': leechers, 'hash': ''}
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: See how we format this from another provider

File: medusa/providers/torrent/html/gimmepeers.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        self.username = None
+        self.password = None
+        self.minseed = None
+        self.minleech = None
+
+        self.cache = tv.Cache(self)
+
+        self.search_params = {
+            # c20=1&c21=1&c25=1&c24=1&c23=1&c22=1&c1=1
+            'c20': 1, 'c21': 1, 'c25': 1, 'c24': 1, 'c23': 1, 'c22': 1, 'c1': 1
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: See how we format this from another provider

File: medusa/providers/torrent/html/gimmepeers.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+        self.urls = {
+            'base_url': 'https://www.gimmepeers.com',
+            'login': 'https://www.gimmepeers.com/takelogin.php',
+            'detail': 'https://www.gimmepeers.com/details.php?id=%s',
+            'search': 'https://www.gimmepeers.com/browse.php',
+            'download': 'https://gimmepeers.com/%s',
+        }
+
+        self.url = self.urls['base_url']
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: Get rid of base_url. See how we do this: https://github.com/pymedusa/Medusa/blob/master/medusa/providers/torrent/html/alpharatio.py#L38-L42

File: medusa/providers/torrent/html/gimmepeers.py (hunk 6 above)
Comment (p0psicles) [reply medusa/providers/torrent/html/gimmepeers.py]: Also check that's file for how we log. We try to move away from the logger.log(). And use like log.debug() log.error() log.exception() etc.

File: medusa/providers/torrent/html/gimmepeers.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+from medusa import logger, tv
+from medusa.bs4_parser import BS4Parser
+from medusa.helper.common import convert_size
+from medusa.providers.torrent.torrent_provider import TorrentProvider
+
+from requests.compat import urljoin
+from requests.utils import dict_from_cookiejar
+
+
```
Comment (p0psicles) [inline medusa/providers/torrent/html/gimmepeers.py]: Youri missing the log import. So this code will fail.

File: medusa/providers/torrent/html/gimmepeers.py (hunk 1 above)
Comment (mystycs) [reply medusa/providers/torrent/html/gimmepeers.py]: i didnt see this in login on other examples. can you show me?

**Comment by mystycs** [comment]
all changes as requested have been implemented @p0psicles @medariox 

thanks!

File: medusa/providers/torrent/html/gimmepeers.py (hunk 6 above)
Comment (mystycs) [reply medusa/providers/torrent/html/gimmepeers.py]: fixed!

File: medusa/providers/torrent/html/gimmepeers.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    def __init__(self):
+
+        TorrentProvider.__init__(self, 'GimmePeers')
+
+        self.url = 'https://www.gimmepeers.com'
+        self.urls = {
+            'login': urljoin(self.url, 'takelogin.php'),
+            'search': urljoin(self.url, 'browse.php'),
+            'download': urljoin(self.url, '%s'),
```
Comment (p0psicles) [inline medusa/providers/torrent/html/gimmepeers.py]: Here you can also use format(). Replace `%s` with `{}`

File: medusa/providers/torrent/html/gimmepeers.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+            if len(torrent_rows) < 2:
+                log.debug(u'Data returned from provider does not contain any torrents')
+                return items
+
+            for result in torrent_rows[1:]:
+                try:
+                    cells = result('td')
+
+                    link = cells[1].find('a')
+                    download_url = self.urls['download'] % cells[2].find('a')['href']
```
Comment (p0psicles) [inline medusa/providers/torrent/html/gimmepeers.py]: Ill correct my previous comment. You can put the urljoin right here. No need to template the string

Comment (mystycs) [reply medusa/providers/torrent/html/gimmepeers.py]: all fixed up, replaced it with urljoin

**Comment by mystycs** [comment]
all fixed up! @p0psicles @medariox

File: medusa/providers/torrent/html/gimmepeers.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+            log.debug(u'Search Mode: {0}'.format(mode))
+            for search_string in search_strings[mode]:
+                if mode != 'RSS':
+                    log.debug(u'Search string: {0}'.format(search_string))
+
+                self.search_params['search'] = search_string
+
+                data = self.session.get(self.urls['search'], params=self.search_params)
+                if not data:
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: please use `if not response or not response.text:` instead

File: medusa/providers/torrent/html/gimmepeers.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+                log.debug(u'Data returned from provider does not contain any torrents')
+                return items
+
+            for result in torrent_rows[1:]:
+                try:
+                    cells = result('td')
+
+                    link = cells[1].find('a')
+                    download_url = urljoin(self.url, cells[2].find('a')['href'])
+                    title = link.getText()
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: replace all getText() with get_text()

File: medusa/providers/torrent/html/gimmepeers.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+                                      title, seeders)
+                        continue
+
+                    item = {
+                        'title': title,
+                        'link': download_url,
+                        'size': size,
+                        'seeders': seeders,
+                        'leechers': leechers,
+                        'hash': '',
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: remove hash

File: medusa/providers/torrent/html/gimmepeers.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+        if re.search('Username or password incorrect!', response.text):
+            log.debug(u'Invalid username or password. Check your settings')
+            return False
+
+        return True
+
+    def _check_auth(self):
+        if not self.username or not self.password:
+            log.debug(u'Invalid username or password. Check your settings')
```
Comment (medariox) [inline medusa/providers/torrent/html/gimmepeers.py]: forgot to add `return False`

**Comment by mystycs** [comment]
@medariox all fixed up. I also fixed up size.

**Comment by mystycs** [comment]
@medariox added the publish date as well! I missed that earlier.

**Pull request merged**